### DISCUSSION TITLE
naming conflicts: a parameter called "metadata" is overridden by the module

### DISCUSSION THREAD

**Discussion by martinkirch** [savonet/liquidsoap#2904] [User Support]
Hello,
When upgrading an old machine to Liquidsoap 2.1.3 we hit an obscure error:
```
    data=metadata.json.stringify(metadata.cover.remove(metadata))

Error 5: this value has type
  'A
but it should be a subtype of
  ('B.{cover : 'C.{remove : 'D}, json : 'E.{stringify : 'F}}) -> 'A
```
This occurs in a function having a single parameter, `metadata`.  It took me some time to guess that this name maybe interferes with the module name. Indeed, changing it to `md` works
````caml
def post_to_showergel(md)
  ignore(http.post(
    headers=[("Content-Type", "application/json")],
    "#{SHOWERGEL}/metadata_log",
    data=metadata.json.stringify(metadata.cover.remove(md))
  ))
end
grille_avec_jingle.on_metadata(fun(m) -> thread.run(fast=false, {post_to_showergel(m)}))
````
:rocket: 

I guess `metadata` is not the only name concerned.

Would it be possible to fail because of the names conflict before the typing error ?

**Comment by smimram** [comment]
Hi Martin,

From the code
```
data=metadata.json.stringify(metadata.cover.remove(metadata))
```
it seems that you are defining a variable `metadata` above. Something like
```
metadata = ...
data=metadata.json.stringify(metadata.cover.remove(metadata))
```
In this case, the variable `metadata` now points to the actual metadata (`...`) instead of the module and Liquidsoap complains that it does not have a method `cover`. So `metadata` is not a reserved keyword, but you should not use this name if you plan to use the`metadata` module afterwards, i.e. do something like
it seems that you are defining a variable `metadata` above. Something like
```
md = ...
data=metadata.json.stringify(metadata.cover.remove(md))
```

Makes sense?

**Comment by martinkirch** [comment]
OK that line was so old I forgot I was also using the name as `metadata.cover.remove`, my bad.

It would be great that such name overrides trigger a warning.
